### ISSUE TITLE
Reproducibility with Embedding layer.

### ISSUE THREAD

**Issue by dbonadiman** [fchollet/keras#1935]
Hello everyone, i want to report (and probably resubmit) an old issue that is in Keras for a while.
I have tried to debug a known issue about reproducibility of the code on GPU.

Many of you have probably noticed that when run on GPU keras code produce slightly different results from time to time. So i tried to debug the problem starting from the the examples of the repository.

The only code that i found not reproducible is the code that involves the Embedding layers.
In particular all the imdb examples are not reproducible. At first i thought it was some problem related to the shuffling of the dataset (in the imdb dataset_loader). But the same lines of code are present in the reuter corpus. And the code is perfectly reproducible. 

Can anybody help me to spot the bug?

PS i was thinking even on some convolution related stuff but on mnist everything is fine.
In personal experiments (work related) with this layer i get the same issue (and it occurs even with a preinitialised word embedding matrix) it may be related to gradient updates? 

Please make sure that the boxes below are checked before you submit your issue. Thank you!
- [x ] Check that you are up-to-date with the master branch of Keras. You can update with:
  pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps
- [x ] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
  pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps
- [ ] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by NasenSpray** [comment]
> PS i was thinking even on some convolution related stuff but on mnist everything is fine.
In personal experiments (work related) with this layer i get the same issue (and it occurs even with a preinitialised word embedding matrix) it may be related to gradient updates?

Are you using cuDNN? cuDNN's backward-pass is non-deterministic by default (at least in [Theano](http://deeplearning.net/software/theano/library/sandbox/cuda/dnn.html)).

**Comment by dbonadiman** [comment]
I know that i fixed that issue but the problem remains even with imdb_lstm.py

**Comment by nouiz** [comment]
There is another op in Theano that can add non-deterministic due to
atomicAdd: https://github.com/Theano/Theano/issues/3029

This issue is to add a flag to allow to get slower, but deterministic
computation.

In the mean time, you can use this diff:

diff --git a/theano/sandbox/cuda/opt.py b/theano/sandbox/cuda/opt.py
index 7a9b953..69d1534 100644
--- a/theano/sandbox/cuda/opt.py
+++ b/theano/sandbox/cuda/opt.py
@@ -1126,7 +1126,7 @@ def local_gpu_advanced_incsubtensor1(node):
                 gpu_op = GpuAdvancedIncSubtensor1(
                     set_instead_of_inc=set_instead_of_inc)
             else:
-                gpu_op = GpuAdvancedIncSubtensor1_dev20(
+                gpu_op = GpuAdvancedIncSubtensor1(
                     set_instead_of_inc=set_instead_of_inc)
             return [gpu_op(as_cuda_ndarray_variable(x),
                            as_cuda_ndarray_variable(y), *coords)]
@@ -1168,7 +1168,7 @@ def local_gpu_advanced_incsubtensor1(node):
                 gpu_op = GpuAdvancedIncSubtensor1(
                     set_instead_of_inc=set_instead_of_inc)
             else:
-                gpu_op = GpuAdvancedIncSubtensor1_dev20(
+                gpu_op = GpuAdvancedIncSubtensor1(
                     set_instead_of_inc=set_instead_of_inc)
             return [host_from_gpu(gpu_op(gpu_x, gpu_y, *coords))]
     return False


On Wed, Mar 9, 2016 at 10:21 AM, Daniele Bonadiman <notifications@github.com

**Comment by dbonadiman** [comment]
Thanks @nouiz, if i'm not wrong this issue should be related to the incsubtensor that is usually used in the embedding layer to  do partial  gradient updates over the whole matrix (i don't remember if Keras does that, but i may assume it is).

**Comment by pinesnow72** [comment]
I just found that trainable(=True) Embedding layer in Keras (with Tensorflow backend) is not reproducible through by experiments with my NER model. When I change to set trainable=False, the result is completely reproducible. Do anyone know why it is not reproducible and how to make it reproducible?
